### PR TITLE
Enhancement/1902 adsense top earning

### PR DESCRIPTION
**Pull request by eugene-manuilov**
## Summary

Addresses issue #1902

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Comment by ryanwelcher** [comment]
@googlebot  I consent.

**Comment by ryanwelcher** [comment]
@felixarntz @aaemnnosttv this looks to be ready for CR.

**Comment by eugene-manuilov** [comment]
@ryanwelcher @aaemnnosttv @felixarntz this one is blocked by #1955. Once that PR is approved and merged, this one will be ready.

File: assets/js/modules/adsense/components/dashboard/DashboardTopEarningPagesWidget.js
Comment (aaemnnosttv) [inline assets/js/modules/adsense/components/dashboard/DashboardTopEarningPagesWidget.js:88]:
```suggestion
	if ( ! Array.isArray( data?.[ 0 ]?.data?.rows ) ) {
```

File: assets/js/modules/adsense/components/dashboard/DashboardTopEarningPagesWidget.js
Comment (felixarntz) [inline assets/js/modules/adsense/components/dashboard/DashboardTopEarningPagesWidget.js:123]: The legacy `Layout` component normally shouldn't be used here, but since this is a data table that has some special formatting ramifications, let's keep it like that for now and cover it via #2022.

File: assets/js/modules/adsense/components/dashboard/DashboardTopEarningPagesWidget.js
Comment (felixarntz) [inline assets/js/modules/adsense/components/dashboard/DashboardTopEarningPagesWidget.js:88]:
```suggestion
	if ( ! data || ! data.length || ! data[ 0 ]?.data?.rows ) {
```
To avoid invalid access of `data[ 0 ]` after.

File: assets/js/modules/adsense/components/dashboard/DashboardTopEarningPagesWidget.js
Comment (felixarntz) [inline assets/js/modules/adsense/components/dashboard/DashboardTopEarningPagesWidget.js:84]: Instead of having this (somewhat inaccurate) logic here, we should rely on the `getAdsenseLinked` selector of the `modules/analytics` store. If that returns `false`, let's return the `AdSenseLinkCTA` here.

Let's omit the `Layout` component here, the CTA itself should be enough.

File: assets/js/modules/adsense/components/dashboard/DashboardTopEarningPagesWidget.js
Comment (felixarntz) [inline assets/js/modules/adsense/components/dashboard/DashboardTopEarningPagesWidget.js:89]: Instead of this new string, let's use `_x( 'Analytics', 'Service name', 'google-site-kit' )` like elsewhere because that is where the data is coming from.

**Review by felixarntz — CHANGES_REQUESTED** [review]
Some initial feedback here.

Since this looks like it's built on top of the PR for #1901 though, I'll continue review in more detail once that is merged.

File: assets/js/modules/adsense/components/dashboard/DashboardTopEarningPagesWidget.js
Comment (felixarntz) [inline assets/js/modules/adsense/components/dashboard/DashboardTopEarningPagesWidget.js:79]: This should use `analytics` I think because the API request is made to Analytics.

File: stories/utils/generate-widget-stories.js
Comment (felixarntz) [inline stories/utils/generate-widget-stories.js:44]: I'd prefer to still keep this called `moduleSlug` just because that's the usual use-case (a single one).

File: stories/utils/generate-widget-stories.js
Comment (felixarntz) [inline stories/utils/generate-widget-stories.js:87]: With those being optional, they should be at the bottom of the list.

File: stories/utils/generate-widget-stories.js
Comment (felixarntz) [inline stories/utils/generate-widget-stories.js:99]: The behavior here feels a bit confusing to me, maybe only because these two are named so similarly but do actually unrelated things. `additionalVariants` adds actual additional variants while `additionalVariantCallbacks` expands already existing variants with extra data. Furthermore I think it's odd that they don't support each other, e.g. I currently cannot provide an `additionalVariantsCallback` for one of the `additionalVariants`.

I suggest the following:
* Make `additionalVariants` a map of `variantName => variantFixtureData` pairs where `variantFixtureData` is an object that has `data` and `options`. The utility should then fill this data the same way as for the automatic existing variants.
* And then, add support for `additionalVariants` to `additionalVariantCallbacks` so that you can also provide extra logic for an additional variant in an additional variant callback.

**Review by felixarntz — CHANGES_REQUESTED** [review]
Mostly LGTM, however I think there's room for iteration on the Storybook utility changes.

File: stories/utils/generate-widget-stories.js
Comment (felixarntz) [inline stories/utils/generate-widget-stories.js:89]: Indentation here should be fixed, all parameter names and parameter descriptions should be aligned.

File: stories/utils/generate-widget-stories.js
Comment (felixarntz) [reply stories/utils/generate-widget-stories.js]: @ryanwelcher I think the change you added here now makes it more complicated than it was before - why change `additionalVariants` to be an array? This makes the code unnecessarily complex. The format of `additionalVariantCallbacks` shouldn't have been changed, and `additionalVariants` should be a map (i.e. object) so that they can be easily looked up by name. Having the name as first value of an array is counter-intuitive I'd say, plus it requires iterating through an array in several places which can be avoided by using an object.

**Review by felixarntz — CHANGES_REQUESTED** [review]
One indentation issue, but more importantly I think the utility for generating widget stories still needs an iteration, it's now more complicated than before IMO.

File: stories/utils/generate-widget-stories.js
Comment (felixarntz) [inline stories/utils/generate-widget-stories.js:179]: If `variantOptions` is an array, we should have here the same type of validation as happens for `options` further above.

File: stories/utils/generate-widget-stories.js
Comment (felixarntz) [inline stories/utils/generate-widget-stories.js:182]: This needs to use `variantOptions`.

File: stories/utils/generate-widget-stories.js
Comment (felixarntz) [inline stories/utils/generate-widget-stories.js:87]: Missed this before, should also be marked as optional as it is.

File: stories/utils/generate-widget-stories.js
Comment (felixarntz) [inline stories/utils/generate-widget-stories.js:190]: Nit-picking, but I think this could be simplified to
```suggestion
			if ( additionalVariantCallbacks[ name ] ) {
				additionalVariantCallbacks[ name ]( dispatch, data, options );
```

File: stories/utils/generate-widget-stories.js
Comment (felixarntz) [inline stories/utils/generate-widget-stories.js:41]: Incorrect indentation here (was overlooked before).

File: stories/module-adsense-components.stories.js
Comment (felixarntz) [inline stories/module-adsense-components.stories.js:34]: This should probably use `Options` instead of `Args` like elsewhere.

**Review by felixarntz — CHANGES_REQUESTED** [review]
Almost LGTM, a few minor improvements to possibly make, only one blocking problem.

**Review by felixarntz — APPROVED** [review]
Great!

**Pull request merged**
